docs/spec/constructors.rst

```diff
@@ -0,0 +1,476 @@
+arguments. If the metaclass is ``type``, this step can be skipped.
+
+If the evaluated return type of the ``__call__`` method is something other than
+``Any`` or an instance of the class being constructed, a type checker should
+assume that the metaclass ``__call__`` method is overriding ``type.__call__``
+in some special manner, and it should not attempt to evaluate the ``__new__``
+or ``__init__`` methods on the class. For example, some metaclass ``__call__``
+methods are annotated to return ``NoReturn`` to indicate that constructor
+calls are not supported for that class.
+
+  ::
+
+    class Meta(type):
+        def __call__(cls, *args, **kwargs) -> NoReturn:
+            raise TypeError("Cannot instantiate class")
+
+    class MyClass(metaclass=Meta):
+        def __new__(cls, *args, **kwargs) -> Self:
+            return super().__new__(cls, *args, **kwargs)
+
+    assert_type(MyClass(), Never)
+
+
+``__new__`` Method
+==================
+
+After the metaclass ``__call__`` method has been evaluated, a type checker
+should evaluate the ``__new__`` method of the class (if applicable) using
+the supplied arguments. This step should be skipped if the class does not
+define a ``__new__`` method and does not inherit a ``__new__`` method from
+a base class other than ``object``.
+
+If the class is generic and explicitly specialized, the type checker should
+partially specialize the ``__new__`` method using the supplied type arguments.
+If the class is not explicitly specialized, class-scoped type variables should
+be solved using the supplied arguments passed to the constructor call.
+
+  ::
+
+    class MyClass[T]:
+        def __new__(cls, x: T) -> Self:
+            return super().__new__(cls)
+
+    # Constructor calls for specialized classes
+    assert_type(MyClass[int](1), MyClass[int])
+    assert_type(MyClass[float](1), MyClass[float])
+    MyClass[int](1.0)  # Type error
+
+    # Constructor calls for non-specialized classes
+    assert_type(MyClass(1), MyClass[int])
+    assert_type(MyClass(1.0), MyClass[float])
+
+If any class-scoped type variables are not solved when evaluating the ``__new__``
+method call using the supplied arguments, these type variables should be left
+unsolved, allowing the ``__init__`` method (if applicable) to be used to solve
+them.
+
+  ::
+
+      class MyClass[T]:
+          def __new__(cls, *args, **kwargs) -> Self:
+              return super().__new__(cls)
+
+          def __init__(self, x: T) -> None:
+              pass
+
+      assert_type(MyClass(1), MyClass[int])
+      assert_type(MyClass(""), MyClass[str])
+
+For most classes, the return type for ``__new__`` method is typically ``Self``,
+but other types are also allowed. For example, the ``__new__`` method may return
+an instance of a subclass or an instance of some completely unrelated class.
+
+If the return type of the ``__new__`` method evaluates to ``Any`` or a union that
```

Comment:
@JelleZijlstra, here's what the latest version has to say about a `__new__` method that returns `Any`.